internal/ingress/controller/controller.go

```diff
@@ -1102,6 +1120,57 @@ func (n *NGINXController) createServers(data []*extensions.Ingress,
 	return servers
 }
 
+// match and merge virtual backends into their respective real backends
+func mergeVirtualBackends(ing *extensions.Ingress, anns *annotations.Ingress, upstreams map[string]*ingress.Backend,
+	servers map[string]*ingress.Server) {
+
+	for _, rule := range ing.Spec.Rules {
+		for _, path := range rule.HTTP.Paths {
+			upsName := fmt.Sprintf("%v-%v-%v",
+				ing.Namespace,
+				path.Backend.ServiceName,
+				path.Backend.ServicePort.String())
+
+			ups := upstreams[upsName]
+
+			if ups.Virtual {
+				merged := false
+
+				// find matching paths
+				for _, server := range servers {
+					for _, location := range server.Locations {
+						if location.Backend == defUpstreamName {
+							continue
+						}
+
+						if location.Path == path.Path && !upstreams[location.Backend].Virtual {
```

Comment:
If I read this correctly you are matching by path only and ignoring server name (host). That means given two apps with same path `/` and different domains respectively myapp.com, yourapp.com and I create canary ingress for my app it will also get applied to yourapp.com

Reply:
I've added a check for host a few lines above this now.